lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,239 @@
+      if (isBuildContext(argType, skipNullable: isGetter)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void check(AstNode node) {
+    // Walk back and look for an async gap that is not guarded by a mounted
+    // property check.
+    AstNode? child = node;
+    while (child != null && child is! FunctionBody) {
+      var parent = child.parent;
+      if (parent is Block) {
+        var statements = parent.statements;
+        var index = statements.indexOf(child as Statement);
+        for (var i = index - 1; i >= 0; i--) {
+          var s = statements.elementAt(i);
+          if (isMountedCheck(s)) {
+            return;
+          } else if (isAsync(s)) {
+            rule.reportLint(node);
+            return;
+          }
+        }
+      } else if (parent is IfStatement) {
+        // if (mounted) { ... do ... }
+        if (isMountedCheck(parent, positiveCheck: true)) {
+          return;
+        }
+      }
+
+      child = parent;
+    }
+  }
+
+  bool isAsync(Statement statement) {
+    var visitor = _AwaitVisitor();
+    statement.accept(visitor);
+    return visitor.hasAwait;
+  }
+
+  /// todo (pq): replace in favor of flutter_utils.isBuildContext
+  bool isBuildContext(DartType? type, {bool skipNullable = false}) {
+    if (type is! InterfaceType) {
+      return false;
+    }
+    if (skipNullable && type.nullabilitySuffix == NullabilitySuffix.question) {
+      return false;
+    }
+    var element = type.element;
+    return element.name == _nameBuildContext &&
+        element.source.uri == _uriFramework;
+  }
+
+  bool isMountedCheck(Statement statement, {bool positiveCheck = false}) {
+    // This is intentionally naive.  Using a simple 'mounted' property check
+    // as a signal plays nicely w/ unanticipated framework classes that provide
+    // their own mounted checks.  The cost of this generality is the possibility
+    // of false negatives.
+    if (statement is IfStatement) {
+      var condition = statement.condition;
+
+      Expression check;
+      if (condition is PrefixExpression) {
+        if (positiveCheck && condition.operator.type == TokenType.BANG) {
+          return false;
+        }
+        check = condition.operand;
+      } else {
+        check = condition;
+      }
+
+      // stateContext.mounted => mounted
+      if (check is PrefixedIdentifier) {
+        check = check.identifier;
+      }
+      if (check is SimpleIdentifier) {
+        if (check.name == 'mounted') {
+          // In the positive case it's sufficient to know we're in a positively
+          // guarded block.
+          if (positiveCheck) {
+            return true;
+          }
+          var then = statement.thenStatement;
+          if (then is ReturnStatement) {
+            return true;
+          }
+          if (then is BreakStatement) {
+            return true;
+          }
+          if (then is Block) {
+            return then.statements.last is ReturnStatement;
```

Comment:
Or `break`? Or `continue`? Consider creating a utility method for "terminates control".

Reply:
Good idea.  Added.